File: lib/commands/services/presences.ts
import { Command } from "commander";
import {
  buildQueries,
  collectQueryValue,
  parseDeprecatedWhereQuery,
  parseFilterQuery,
} from "../utils/query.js";
import { sdkForProject } from "../../sdks.js";
import {
  actionRunner,
  commandDescriptions,
  success,
  parse,
  parseBool,
  parseInteger,
  parseJsonObject,
} from "../../parser.js";
import { Presences } from "@appwrite.io/console";

let presencesClient: Presences | null = null;

const getPresencesClient = async (): Promise<Presences> => {
  if (!presencesClient) {
    const sdkClient = await sdkForProject();
    presencesClient = new Presences(sdkClient);
  }
  return presencesClient;
};

export const presences = new Command("presences")
  .description(commandDescriptions["presences"] || "The Presences service allows you to track and manage real-time user presence in your project.")
  .configureHelp({
    helpWidth: process.stdout.columns || 80,
  });

const presencesListCommand = presences
  .command(`list`)
  .description(`List presence logs. Expired entries are filtered out automatically.
`)
  .option(`--queries [queries...]`, `Raw Appwrite JSON query strings (legacy). Use this for advanced queries or automation; for common filtering, sorting, and pagination prefer --filter, --sort-asc, --sort-desc, --limit, and --offset. When mixed, raw --queries are sent before generated flag queries. Array of query strings generated using the Query class provided by the SDK.`)
  .option(
    `--total [value]`,
    `When set to false, the total count returned will be 0 and will not be calculated.`,
    (value: string | undefined) =>
      value === undefined ? true : parseBool(value),
  )
  .option(`--ttl <ttl>`, `TTL (seconds) for caching list responses. Responses are stored in an in-memory key-value cache, keyed per project, collection, schema version (attributes and indexes), caller authorization roles, and the exact query — so users with different permissions never share cached entries. Schema changes invalidate cached entries automatically; document writes do not, so choose a TTL you are comfortable serving as stale data. Set to 0 to disable caching. Must be between 0 and 86400 (24 hours).`, parseInteger)
  .option(`--filter <expression>`, `Filter using a simple comparison expression. Repeat for multiple filters. Supports field=value, field!=value, field>value, field>=value, field<value, and field<=value.`, (value: string, previous: string[] | undefined) => collectQueryValue(parseFilterQuery(value), previous))
  .option(`--where <expression>`, `Deprecated. Use --filter instead. Filter using a simple comparison expression. Repeat for multiple filters.`, (value: string, previous: string[] | undefined) => collectQueryValue(parseDeprecatedWhereQuery(value), previous))
  .option(`--sort-asc <attribute>`, `Sort results by an attribute in ascending order. Repeat for multiple sort fields.`, (value: string, previous: string[] | undefined) => collectQueryValue(value, previous))
  .option(`--sort-desc <attribute>`, `Sort results by an attribute in descending order. Repeat for multiple sort fields.`, (value: string, previous: string[] | undefined) => collectQueryValue(value, previous))
  .option(`--limit <limit>`, `Maximum number of results to return.`, parseInteger)
  .option(`--offset <offset>`, `Number of results to skip.`, parseInteger)
  .option(`--cursor-after <id>`, `Return results after this cursor ID.`)
  .option(`--cursor-before <id>`, `Return results before this cursor ID.`)
  .action(
    actionRunner(
      async ({ queries, total, ttl, filter, where, sortAsc, sortDesc, cursorAfter, cursorBefore, limit, offset }) =>
        parse(await (await getPresencesClient()).list(buildQueries({ queries, filter, where, sortAsc, sortDesc, cursorAfter, cursorBefore, limit, offset }), total, ttl)),
    ),
  );


const presencesGetUsageCommand = presences
  .command(`get-usage`)
  .description(`Get presence usage metrics, including the current total of online users and historical online user counts for the selected time range.
`)
  .option(`--range <range>`, `Date range.`)
  .action(
    actionRunner(
      async ({ range }) =>
        parse(await (await getPresencesClient()).getUsage(range)),
    ),
  );


const presencesGetCommand = presences
  .command(`get`)
  .description(`Get a presence log by its unique ID. Entries whose \`expiresAt\` is in the past are treated as not found.
`)
  .requiredOption(`--presence-id <presence-id>`, `Presence unique ID.`)
  .action(
    actionRunner(
      async ({ presenceId }) =>
        parse(await (await getPresencesClient()).get(presenceId)),
    ),
  );


const presencesUpsertCommand = presences
  .command(`upsert`)
  .description(`Create or update a presence log by its user ID.
`)
  .requiredOption(`--presence-id <presence-id>`, `Presence unique ID.`)
  .requiredOption(`--status <status>`, `Presence status.`)
  .option(`--permissions [permissions...]`, `An array of permissions strings. By default, only the current user is granted all permissions. Learn more about permissions (https://appwrite.io/docs/permissions).`)
  .option(`--expires-at <expires-at>`, `Presence expiry datetime.`)
  .option(`--metadata <metadata>`, `Presence metadata object.`)
  .action(
    actionRunner(
      async ({ presenceId, status, permissions, expiresAt, metadata }) =>
        parse(await (await getPresencesClient()).upsert(presenceId, status, permissions, expiresAt, parseJsonObject(metadata, "--metadata"))),
    ),
  );


const presencesUpdateCommand = presences
  .command(`update`)
  .description(`Update a presence log by its unique ID. Using the patch method you can pass only specific fields that will get updated.
`)
  .requiredOption(`--presence-id <presence-id>`, `Presence unique ID.`)
  .option(`--status <status>`, `Presence status.`)
  .option(`--expires-at <expires-at>`, `Presence expiry datetime.`)
  .option(`--metadata <metadata>`, `Presence metadata object.`)
  .option(`--permissions [permissions...]`, `An array of permissions strings. By default, only the current user is granted all permissions. Learn more about permissions (https://appwrite.io/docs/permissions).`)
  .option(
    `--purge [value]`,
    `When true, purge cached responses used by list presences endpoint.`,
    (value: string | undefined) =>
      value === undefined ? true : parseBool(value),
  )
  .action(
    actionRunner(
      async ({ presenceId, status, expiresAt, metadata, permissions, purge }) =>
        parse(await (await getPresencesClient()).update(presenceId, status, expiresAt, parseJsonObject(metadata, "--metadata"), permissions, purge)),
    ),
  );


const presencesDeleteCommand = presences
  .command(`delete`)
  .description(`Delete a presence log by its unique ID.
`)
  .requiredOption(`--presence-id <presence-id>`, `Presence unique ID.`)
  .action(
    actionRunner(
      async ({ presenceId }) =>
        parse(await (await getPresencesClient()).delete(presenceId)),
    ),
  );

File: internal/cmd/services/presences.go
package services

import (
	"github.com/spf13/cobra"

	"github.com/appwrite/sdk-for-go/v6/presences"

	"github.com/appwrite/sdk-for-cli/internal/app"
	"github.com/appwrite/sdk-for-cli/internal/query"
	"github.com/appwrite/sdk-for-cli/internal/sdk"
)

// NewPresencesCommand builds the `presences` command tree.
func NewPresencesCommand() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "presences",
		Short: "The Presences service allows you to track and manage real-time user presence in your project.",
	}

	cmd.AddCommand(newPresencesListCommand())
	cmd.AddCommand(newPresencesGetCommand())
	cmd.AddCommand(newPresencesDeleteCommand())

	return cmd
}

func newPresencesListCommand() *cobra.Command {
	var queries []string
	var total bool
	var ttl int
	var filter []string
	var where []string
	var sortAsc []string
	var sortDesc []string
	var limit int
	var offset int
	var cursorAfter string
	var cursorBefore string

	cmd := &cobra.Command{
		Use:   "list",
		Short: "List presence logs. Expired entries are filtered out automatically.\n",
		RunE: func(cmd *cobra.Command, args []string) error {
			client, err := app.ClientForProject("")
			if err != nil {
				return err
			}
			service := presences.New(client)

			parsedFilter, err := query.ParseFilters(filter)
			if err != nil {
				return err
			}
			parsedWhere, err := query.ParseFilters(where)
			if err != nil {
				return err
			}

			queries, err := query.Build(query.Options{
				Queries:      queries,
				Filter:       parsedFilter,
				Where:        parsedWhere,
				SortAsc:      sortAsc,
				SortDesc:     sortDesc,
				Limit:        app.FlagInt(cmd, "limit", limit),
				Offset:       app.FlagInt(cmd, "offset", offset),
				CursorAfter:  app.FlagString(cmd, "cursor-after", cursorAfter),
				CursorBefore: app.FlagString(cmd, "cursor-before", cursorBefore),
			})
			if err != nil {
				return err
			}

			// An unset flag must be omitted, not sent as its zero value: the
			// TypeScript passes undefined and the SDK drops it.
			options := []presences.ListOption{}
			if cmd.Flags().Changed("queries") {
				options = append(options, service.WithListQueries(queries))
			}
			if cmd.Flags().Changed("total") {
				options = append(options, service.WithListTotal(total))
			}
			if cmd.Flags().Changed("ttl") {
				options = append(options, service.WithListTtl(ttl))
			}

			result, err := service.List(options...)
			if err != nil {
				return sdk.WrapMutationError("GET", err)
			}

			return app.Render(result)
		},
	}

	cmd.Flags().StringArrayVar(&queries, "queries", nil, "Array of query strings generated using the Query class provided by the SDK.")
	cmd.Flags().BoolVar(&total, "total", false, "When set to false, the total count returned will be 0 and will not be calculated.")
	cmd.Flags().Lookup("total").NoOptDefVal = "true"
	cmd.Flags().IntVar(&ttl, "ttl", 0, "TTL (seconds) for caching list responses. Responses are stored in an in-memory key-value cache, keyed per project, collection, schema version (attributes and indexes), caller authorization roles, and the exact query — so users with different permissions never share cached entries. Schema changes invalidate cached entries automatically; document writes do not, so choose a TTL you are comfortable serving as stale data. Set to 0 to disable caching. Must be between 0 and 86400 (24 hours).")
	cmd.Flags().StringArrayVar(&filter, "filter", nil, "Filter using a simple comparison expression. Repeat for multiple filters. Supports field=value, field!=value, field>value, field>=value, field<value, and field<=value.")
	cmd.Flags().StringArrayVar(&where, "where", nil, "Deprecated. Use --filter instead. Filter using a simple comparison expression. Repeat for multiple filters.")
	cmd.Flags().StringArrayVar(&sortAsc, "sort-asc", nil, "Sort results by an attribute in ascending order. Repeat for multiple sort fields.")
	cmd.Flags().StringArrayVar(&sortDesc, "sort-desc", nil, "Sort results by an attribute in descending order. Repeat for multiple sort fields.")
	cmd.Flags().IntVar(&limit, "limit", 0, "Maximum number of results to return.")
	cmd.Flags().IntVar(&offset, "offset", 0, "Number of results to skip.")
	cmd.Flags().StringVar(&cursorAfter, "cursor-after", "", "Return results after this cursor ID.")
	cmd.Flags().StringVar(&cursorBefore, "cursor-before", "", "Return results before this cursor ID.")
	return cmd
}

func newPresencesGetCommand() *cobra.Command {
	var presenceId string

	cmd := &cobra.Command{
		Use:   "get",
		Short: "Get a presence log by its unique ID. Entries whose `expiresAt` is in the past are treated as not found.\n",
		RunE: func(cmd *cobra.Command, args []string) error {
			client, err := app.ClientForProject("")
			if err != nil {
				return err
			}
			service := presences.New(client)

			result, err := service.Get(presenceId)
			if err != nil {
				return sdk.WrapMutationError("GET", err)
			}

			return app.Render(result)
		},
	}

	cmd.Flags().StringVar(&presenceId, "presence-id", "", "Presence unique ID.")
	_ = cmd.MarkFlagRequired("presence-id")
	return cmd
}

func newPresencesDeleteCommand() *cobra.Command {
	var presenceId string

	cmd := &cobra.Command{
		Use:   "delete",
		Short: "Delete a presence log by its unique ID.\n",
		RunE: func(cmd *cobra.Command, args []string) error {
			client, err := app.ClientForProject("")
			if err != nil {
				return err
			}
			service := presences.New(client)

			result, err := service.Delete(presenceId)
			if err != nil {
				return sdk.WrapMutationError("DELETE", err)
			}

			return app.Render(result)
		},
	}

	cmd.Flags().StringVar(&presenceId, "presence-id", "", "Presence unique ID.")
	_ = cmd.MarkFlagRequired("presence-id")
	return cmd
}
